### PR TITLE
machine/rp2040: implement i2c

### PR DESCRIPTION
**Pull request by soypat**
Please take a look. Tested it with mpu6050 driver. Solves #1989.

### CODE REVIEW

File: src/machine/machine_rp2040_i2c.go [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+	SDA, SCL Pin
+}
+
+type I2C struct {
+	Bus           *rp.I2C0_Type
+	restartOnNext bool
+}
+
+var (
+	errInvalidI2CBaudrate = errors.New("invalid i2c baudrate")
```
Comment (deadprogram) [inline src/machine/machine_rp2040_i2c.go]: Please export errors here.

File: src/machine/machine_rp2040_i2c.go [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+
+// Configure initializes i2c peripheral and configures I2C config's pins passed.
+// If I2CConfig pins are not set then the default I2C0 bus is used on SDA,SCL=GPIO4,GPIO5
+// and SDA,SCL=GPIO2,GPIO3 for I2C1.
+func (i2c *I2C) Configure(config I2CConfig) error {
+	if config.SCL == 0 {
+		// Default I2C0 port selected
+		switch i2c.Bus {
+		case rp.I2C0:
+			config.SDA = GPIO4
```
Comment (deadprogram) [inline src/machine/machine_rp2040_i2c.go]: See my comments about overriding pins in the SPI PR.

File: src/machine/machine_rp2040_i2c.go [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+	i2c.Bus.IC_FS_SPKLEN.Set(u32max(1, lcnt/16))
+
+	i2c.Bus.IC_SDA_HOLD.ReplaceBits(sdaTxHoldCnt<<rp.I2C0_IC_SDA_HOLD_IC_SDA_TX_HOLD_Pos, rp.I2C0_IC_SDA_HOLD_IC_SDA_TX_HOLD_Msk, 0)
+	i2c.enable()
+	return nil
+}
+
+//go:inline
+func (i2c *I2C) enable() {
+	// i2c.Bus.IC_ENABLE.Set(1)
```
Comment (deadprogram) [inline src/machine/machine_rp2040_i2c.go]: Should this be removed.

File: targets/rp2040.json [hunk 4]
```diff
@@ -7,6 +7,7 @@
     "binary-format": "uf2",
     "uf2-family-id": "0xe48bff56",
     "rp2040-boot-patch": true,
+    "openocd-transport":"rp2040",
```
Comment (deadprogram) [inline targets/rp2040.json]: It is this actually to do with this specific PR? Seems like something else.

**Comment by deadprogram** [comment]
The I2C pin definitions are needed here just like in the SPI PR, please.

File: targets/rp2040.json (hunk 4 above)
Comment (soypat) [reply targets/rp2040.json]: Yeah sorry, this was me testing picoprobe's debugging ability,

**Comment by deadprogram** [comment]
Hello @soypat now that #1991 has been merged there are some conflicts here that need to be addressed, if you please.

**Comment by soypat** [comment]
@deadprogram had to rollback merge to squash commits. Should be A-OK now.

**Comment by deadprogram** [comment]
Tested with my SSD1306 display and worked perfectly. Thanks @soypat now squash/merging.

**Pull request merged**